docs/hugo/content/design/ADR-2022-09-Reading-Status-Properties-Of-Resources.md

```diff
@@ -0,0 +1,261 @@
+  location: Germany West Central
+  owner:
+    name: dev-sample-rg
+---
+apiVersion: authorization.azure.com/v1beta20200801preview
+kind: RoleAssignment
+metadata:
+  name: 6a2d44f5-57d8-4916-9f46-ff7c9c1b338f
+spec:
+  location: Germany West Central
+  owner:
+    name: samplevnet
+    group: network.azure.com
+    kind: VirtualNetwork
+  principalId:
+    ref:
+      name: sample-uai
+      group: manangedidentity.azure.com
+      kind: UserAssignedIdentity
+      path: status.principalId
+  roleDefinitionReference:
+    armId: /subscriptions/00000000-0000-0000-0000-000000000000/providers/Microsoft.Authorization/roleDefinitions/b24988ac-6180-42a0-ab88-20f7382dd24c
+```
+
+Pros:
+
+1. More powerful. Could be reused in the future if/when more of this sort of requirement comes up.
+
+Cons:
+
+1. More verbose for users. At least in this specific example it's a lot more writing (and possibility for mistakes) than the shorter alternatives.
+2. Errors are likely to be harder to decipher. Without something in status that says what this actually resolved to, it might be hard to figure out what exactly you sent to Azure (unless Azure specifically calls it out in the error response that shows up in the Ready Condition. There's no guarantees that all RPs do this). We could mitigate this by putting a section in `status` that has what we resolved this to, although doing that will complicate the reconcile pass. Alternatively we could punt on that and see how difficult users find it. Most Azure services probably give details about what value was sent, and we could error on empty-string to protect against typos.
+3. Risk for privilege escalation, where a user in the namespace that doesn't have permission to read resources of type `X` instead asks ASO to read that resource. Azure then reflects that value back to them in the error message.
+4. Uncertain general reconcile behavior. This reference constitutes a dependency on the referenced resource. Does this mean that ideally, the resource holding the reference will be `Watch`-ing the referenced resource and immediately react to changes in its dependencies? Implementing that dynamic watch behavior given the current structure of controller-runtime would be very difficult.
+
+### Option 3: Support reading/writing to a ConfigMap or Secret, and indirect through that
+
+The idea here is that `UserManagedIdentity` would allow exporting a Secret or ConfigMap value, and then the `RoleAssignment` resource would support (optionally) sourcing from that store instead of having the value hardcoded.
+The main difference between this and what we have now is for references we can't replace the entire reference with a `SecretReference`, we need a new type like `OptionalSecretReference`:
+
+We would only opt a select few resource properties into this `OptionalSecretReference` structure. We could (via serialization magic) support new properties adopting this shape in a nonbreaking way, in case there are properties we didn't think
```

Comment:
**Task** does this kind of serialization magic in a number of places - one example is around [variables](https://taskfile.dev/usage/#variables) where they can be directly defined:

```
    vars:
      VAR: Hello!
```
or dynamically defined:
```
    vars:
      GIT_COMMIT:
        sh: git log -n 1 --format=%h
```

We can reference the [Task implementation](https://github.com/go-task/task/blob/master/taskfile/var.go) for an approach.